components/calendar/src/gregorian.rs

```diff
@@ -233,3 +233,197 @@ pub(crate) fn year_as_gregorian(year: i32) -> types::FormattableYear {
+                iso_month: 9,
+                iso_day: 5,
+                expected_year: 4,
+                expected_era: Era(tinystr!(16, "ce")),
+                expected_month: 9,
+                expected_day: 5,
+            },
+            TestCase {
+                fixed_date: 36219,
+                iso_year: 100,
+                iso_month: 3,
+                iso_day: 1,
+                expected_year: 100,
+                expected_era: Era(tinystr!(16, "ce")),
+                expected_month: 3,
+                expected_day: 1,
+            },
+        ];
+
+        for case in cases {
+            check_test_case(case);
+        }
+    }
+
+    #[test]
+    fn test_gregorian_bce() {
+        // Tests that the Gregorian calendar gives the correct expected
+        // day, month, and year for negative years (BC/BCE/pre-gregory era)
+
+        let cases = [
+            TestCase {
+                fixed_date: 0,
+                iso_year: 0,
+                iso_month: 12,
+                iso_day: 31,
+                expected_year: 1,
+                expected_era: Era(tinystr!(16, "bce")),
+                expected_month: 12,
+                expected_day: 31,
+            },
+            TestCase {
+                fixed_date: -365, // This is a leap year
+                iso_year: 0,
+                iso_month: 1,
+                iso_day: 1,
+                expected_year: 1,
+                expected_era: Era(tinystr!(16, "bce")),
+                expected_month: 1,
+                expected_day: 1,
+            },
+            TestCase {
+                fixed_date: -366,
+                iso_year: -1,
+                iso_month: 12,
+                iso_day: 31,
+                expected_year: 2,
+                expected_era: Era(tinystr!(16, "bce")),
+                expected_month: 12,
+                expected_day: 31,
+            },
+            TestCase {
+                fixed_date: -1461,
+                iso_year: -4,
+                iso_month: 12,
+                iso_day: 31,
+                expected_year: 5,
+                expected_era: Era(tinystr!(16, "bce")),
+                expected_month: 12,
+                expected_day: 31,
+            },
+            TestCase {
+                fixed_date: -1826,
+                iso_year: -4,
+                iso_month: 1,
+                iso_day: 1,
+                expected_year: 5,
+                expected_era: Era(tinystr!(16, "bce")),
+                expected_month: 1,
+                expected_day: 1,
+            },
+        ];
+
+        for case in cases {
+            check_test_case(case);
+        }
+    }
+
+    #[test]
+    fn check_gregorian_directionality() {
+        // Tests that for a large range of fixed dates, if a fixed date
+        // is less than another, the corresponding YMD should also be less
+        // than the other, without exception.
+        for i in -100..100 {
+            for j in -100..100 {
```

Comment:
Thought: Another coverage option would be `i in -1000..1000` and `k in -10..10` with `j = i + k`. Not sure which is better; I think the test gets the job done either way. We should eventually cover a larger range via fuzz testing.